### DISCUSSION TITLE
Data filter on Cup Category

### DISCUSSION THREAD

**Discussion by Maaaarden** [kaiusl/KLPlugins.DynLeaderboards#24] [Ideas]
I'm driving a league that is using the Cup Category as a separator for different groups of drivers.
It would be great to be able to have a dashboard showing only 'Pro'/'ProAm'/'Am' categories.

**Comment by kaiusl** [comment]
Would a leaderboard that only shows the players class cars with the players cup category suffice? 

I'm thinking that we could add Cup, RelativeCup and PartialRelativeCup leaderboard types which are like Class variants but only show the cars in the players Cup Category. It should be relatively quick and easy to add and I'm happy to do it.

A more complete and more flexible solution would be to add some kind proper filtering system which would work with any leaderboard type. I quite like this idea but as far as I can see it would require quite a bit of extra work that I'm not really interested in taking up at this moment.

**Reply by Maaaarden** [reply]
Yeah, Cup, RelativeCup and PartialRelativeCup would suffice.

I'm trying to study the code to see if i could assist in anything - but im struggling to figure out where this is done :)

**Reply by kaiusl** [reply]
Yeah it's maybe a bit of a mess :/ I have been meaning to do clean up and add bunch of documentations but never got around to it unfortunately.

However a quick overview how it work is that [`Values`] class hold all the data. It has `Cars` property which holds all the cars in the session ordered by their overall position. Each leaderboard type then has a list of indices into this `Cars` list and that's how their order is set. For example `PosInClassCarsIdxs` property hold the indices for the players class cars. The `LeaderboardValues` property hold different leaderboards the user has created and inside it are the relative leaderboards car indices (for example [`_relativePosOnTrackCarsIdxs`]). It also has bunch of delegates which change based on which leaderboard is selected. These indices into `Cars` are updated it [`Values.UpdateCarData`] method. So we'll need to mirror what's done with Class related indices.

[`Values`]: https://github.com/kaiusl/KLPlugins.DynLeaderboards/blob/9350e68a2b01f243be8a86a48956e5ef15622e39/src/components/Values.cs#L36C20-L36C26
[`_relativePosOnTrackCarsIdxs`]: https://github.com/kaiusl/KLPlugins.DynLeaderboards/blob/9350e68a2b01f243be8a86a48956e5ef15622e39/src/components/DynLeaderboardValues.cs#L30C25-L30C52
 [`Values.UpdateCarData`]: https://github.com/kaiusl/KLPlugins.DynLeaderboards/blob/9350e68a2b01f243be8a86a48956e5ef15622e39/src/components/Values.cs#L547

**Reply by Maaaarden** [reply]
I dont believe its your code that is the issue. More the lack of skills on my hands.

Think ill have some time to look into this later today to see if i can figure it out 


Nice work btw. Really useful dashboards:)

**Reply by kaiusl** [reply]
Thank you.

I started to add these leaderboards in #25. Basic leaderboards work but they are missing bunch of properties like gaps to cup leaders and similar. If you are interested in what needs to change you can see it there.

**Reply by Maaaarden** [reply]
Great progress already, eh? 

I'm not having as much time on my hands as i would like to - but look pretty simple, the stuff i've currently looked into that you've made of changes.

**Reply by kaiusl** [reply]
> Great progress already, eh?

Yeah. Everything needed should be there. Need to do a bit more testing now.

**Reply by Maaaarden** [reply]
Anything I can do to help testing this?

**Reply by kaiusl** [reply]
You could give it a try and see if everything looks ok. I don't think there is anything special to look for. You'll also need to enable the new leaderboard types from the plugin setting, they are disabled by default on existing configurations.

I uploaded a new [preview version](https://github.com/kaiusl/KLPlugins.DynLeaderboards/releases) so you could easily try it.

**Reply by Maaaarden** [reply]
I'm not home until tomorrow - but then i actually have a race in my league that I can test this with.

**Reply by Maaaarden** [reply]
Just tried to download - but it says the zip is empty on beta2 zip

**Reply by kaiusl** [reply]
I made a new archive, give it another try ;)

**Reply by Maaaarden** [reply]
A quick test on a singleplayer race with multiple cup present, seems like its working like a charm. Couldn't do laps or anything since its on a laptop. 
Looking forward to a proper test tomorrow :)

**Reply by Maaaarden** [reply]
Race finished - great leaderboard functionaly to have at hand - really appreciate the work effort :)
